File: python/257_BinaryTreePaths.py
from unittest import TestCase
from Definitions import TreeNode
# https://leetcode.com/problems/binary-tree-paths/


class BinaryTreePaths(TestCase):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        ret = []

        def dfs(node, path):
            if not node: return
            path = path + [node.val]
            if node.left is None and node.right is None:
                ret.append("->".join(map(str, path)))
            dfs(node.left, path)
            dfs(node.right, path)
        dfs(root, [])
        return ret

    def test1(self):
        root = TreeNode(1)
        root.left = TreeNode(2)
        root.left.right = TreeNode(5)
        root.right = TreeNode(3)
        self.assertEqual(["1->2->5", "1->3"], self.binaryTreePaths(root))

File: kotlin/src/main/kotlin/257.BinaryTreePaths.kt
import org.junit.Assert.assertEquals
import org.junit.Test
//https://leetcode.com/problems/binary-tree-paths/


class BinaryTreePaths{
  fun binaryTreePaths(root: TreeNode?): List<String> {
    val ret = mutableListOf<String>()
    fun dfs(node: TreeNode?, path: List<Int>){
      if(node == null) return

      val p = path + listOf(node.`val`)
      if(node.left == null && node.right == null)
        ret.add(p.joinToString("->"))
      dfs(node.left, p)
      dfs(node.right, p)
    }
    dfs(root, emptyList())
    return ret
  }

  @Test
  fun test1(){
    val root = TreeNode(1).also{
      it.left = TreeNode(2).also { it.right = TreeNode(5) }
      it.right = TreeNode(3)
    }
    assertEquals(listOf("1->2->5", "1->3"), binaryTreePaths(root))
  }

  @Test
  fun test2(){
    assertEquals(listOf<String>(), binaryTreePaths(null))
  }
}